### ISSUE TITLE
app.py

### ISSUE THREAD

**Issue by Afjal9** [Afjal9/judgment-app#1]
# app.py
import io
import re
import math
import hashlib
from typing import List, Tuple, Dict, Any

import streamlit as st

# --- PDF parsing ---
from pdfminer.high_level import extract_pages
from pdfminer.layout import LTTextContainer, LTChar, LTTextLineHorizontal

# --- NLP / search ---
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Optional: light, local sentence embedding (no internet needed after first install)
# Using sentence-transformers for semantic search quality
try:
    from sentence_transformers import SentenceTransformer
    _HAS_ST = True
except Exception:
    _HAS_ST = False

# Basic sentence splitter (rule-based; avoids needing NLTK download)
_SENT_SPLIT = re.compile(r'(?<!\b[A-Z])[.?!]\s+(?=[A-Z(])|(?<=\n)\s*')

# ------------------------
# Utilities
# ------------------------

def normalize_ws(s: str) -> str:
    return re.sub(r'\s+', ' ', s).strip()

def split_sentences(text: str) -> List[str]:
    # Conservative split; keeps initials/abbrev safer than naive ". "
    raw = [t.strip() for t in _SENT_SPLIT.split(text) if t.strip()]
    # Merge very short fragments with following sentence
    out = []
    buff = ""
    for t in raw:
        if len(t) < 30:
            buff = (buff + " " + t).strip()
            continue
        if buff:
            out.append((buff + " " + t).strip())
            buff = ""
        else:
            out.append(t)
    if buff:
        out.append(buff)
    return out

def page_text_lines(pdf_bytes: bytes) -> List[List[str]]:
    """
    Extract text page-by-page and keep line boundaries to report page/line numbers.
    Uses pdfminer to walk layout and collect line texts reliably.
    Returns: list of pages, each page is list of line strings
    """
    pages_lines: List[List[str]] = []
    for page_layout in extract_pages(io.BytesIO(pdf_bytes)):
        lines = []
        for element in page_layout:
            if isinstance(element, LTTextContainer):
                for text_line in element:
                    if isinstance(text_line, LTTextLineHorizontal):
                        line = text_line.get_text()
                        # Clean control characters, trailing newlines
                        line = line.replace('\x0c', '').rstrip()
                        if line.strip():
                            lines.append(normalize_ws(line))
        # Preserve visual order: pdfminer yields in reading order
        pages_lines.append(lines)
    return pages_lines

def build_indices(pages_lines: List[List[str]]) -> Dict[str, Any]:
    """
    Build:
      - full_text
      - sentence-level index with (page, line_start, line_end)
      - line-level index with (page, line_no)
    """
    line_index = []  # [(page, line_no, text)]
    for p, lines in enumerate(pages_lines, start=1):
        for i, ln in enumerate(lines, start=1):
            if ln.strip():
                line_index.append((p, i, ln))

    # Concatenate per page to do sentence splitting while tracking approx line span
    sentence_index = []  # [(sentence_text, page, line_start, line_end)]
    for p, lines in enumerate(pages_lines, start=1):
        if not lines:
            continue
        text = "\n".join(lines)
        sents = split_sentences(text)
        # Map sentences back to line spans by greedy matching
        # Build a rolling pointer over lines
        joined_lines = "\n".join(lines)
        cursor = 0
        # Precompute line char offsets
        offsets = []
        off = 0
        for ln in lines:
            offsets.append((off, off + len(ln)))
            off += len(ln) + 1  # +1 for "\n"

        def charpos_to_line(cp: int) -> int:
            # binary search over offsets
            lo, hi = 0, len(offsets)-1
            ans = 0
            while lo <= hi:
                mid = (lo + hi)//2
                s,e = offsets[mid]
                if s <= cp <= e:
                    return mid + 1
                if cp < s: hi = mid - 1
                else: lo = mid + 1
            return max(1, min(len(offsets), lo+1))

        for s in sents:
            s_norm = normalize_ws(s)
            if not s_norm:
                continue
            # find s in joined_lines starting at cursor
            idx = joined_lines.find(s_norm, cursor)
            if idx == -1:
                # fallback: search anywhere
                idx = joined_lines.find(s_norm)
            if idx == -1:
                # give up; still include sentence with unknown lines
                sentence_index.append((s_norm, p, 1, max(1, len(lines))))
                continue
            start_cp = idx
            end_cp = idx + len(s_norm)
            ls = charpos_to_line(start_cp)
            le = charpos_to_line(end_cp)
            sentence_index.append((s_norm, p, ls, le))
            cursor = end_cp

    full_text = "\n".join(["\n".join(lines) for lines in pages_lines])
    return {
        "line_index": line_index,
        "sentence_index": sentence_index,
        "full_text": full_text
    }

def embed_texts(texts: List[str], model=None) -> np.ndarray:
    if model is not None:
        emb = np.array(model.encode(texts, normalize_embeddings=True))
    else:
        # TF-IDF fallback if SentenceTransformer isn't available
        vec = TfidfVectorizer(ngram_range=(1,2), min_df=1).fit(texts)
        emb = vec.transform(texts)  # sparse
        # Wrap to return (vectorizer, matrix)
        return (vec, emb)
    return emb

def embed_query(q: str, model=None, tfidf_pack=None):
    if model is not None:
        return np.array(model.encode([q], normalize_embeddings=True))
    else:
        vec, mat = tfidf_pack
        return vec.transform([q])

def top_k(sim: np.ndarray, k=5) -> List[int]:
    k = min(k, sim.shape[1] if sim.ndim == 2 else sim.shape[0])
    idx = np.argpartition(-sim, range(k))[:k]
    # sort those top-k by score desc
    idx = idx[np.argsort(-sim[idx])]
    return idx.tolist()

def extractive_summary(sentences: List[str], max_sents=6) -> str:
    """
    Simple unsupervised summary using TF-IDF + cosine centrality (TextRank-ish).
    """
    if not sentences:
        return ""
    # Limit to avoid memory blow-up
    sents = [normalize_ws(s) for s in sentences if len(s.strip()) > 0][:1200]
    vec = TfidfVectorizer(ngram_range=(1,2), min_df=1, stop_words='english')
    X = vec.fit_transform(sents)
    sim = (X * X.T).toarray()
    centrality = sim.sum(axis=1)
    order = np.argsort(-centrality)[:max_sents]
    # Sort selected sentences by original order for readability
    order = sorted(order)
    chosen = [sents[i] for i in order]
    # Merge to 1–2 paragraphs
    mid = math.ceil(len(chosen)/2)
    p1 = " ".join(chosen[:mid])
    p2 = " ".join(chosen[mid:])
    return (p1 + ("\n\n" + p2 if p2 else "")).strip()

# ------------------------
# Streamlit App
# ------------------------

st.set_page_config(page_title="Judgment Insight", page_icon="⚖️", layout="wide")
st.title("⚖️ Judgment Insight — Upload • Ask • Locate • Summarize")

with st.sidebar:
    st.markdown("### Settings")
    st.markdown("- Uses local NLP (no internet).")
    use_st = st.checkbox("Use Semantic Embeddings (better results)", value=True)
    st.caption("If unchecked, falls back to TF-IDF.")

# Model load (once)
@st.cache_resource(show_spinner=False)
def load_semantic_model(enabled=True):
    if enabled and _HAS_ST:
        try:
            return SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")
        except Exception:
            return None
    return None

model = load_semantic_model(use_st)

uploaded = st.file_uploader("Upload a PDF judgment", type=["pdf"])

if uploaded is None:
    st.info("Upload a PDF to begin. The app will index pages, lines, and sentences for precise answers.")
    st.stop()

pdf_bytes = uploaded.read()

@st.cache_data(show_spinner=True)
def parse_and_index(pdf_bytes: bytes):
    pages_lines = page_text_lines(pdf_bytes)
    idx = build_indices(pages_lines)
    # Build embeddings
    sentences = [s for (s, p, ls, le) in idx["sentence_index"]]
    lines = [t for (_, _, t) in idx["line_index"]]

    # Sentence embeddings
    if model is not None:
        sent_emb = embed_texts(sentences, model=model)
        line_emb = embed_texts(lines, model=model)
        tfidf_pack_sent = None
        tfidf_pack_line = None
    else:
        tfidf_pack_sent = embed_texts(sentences, model=None)  # (vec, mat)
        tfidf_pack_line  = embed_texts(lines, model=None)
        sent_emb = tfidf_pack_sent[1]
        line_emb = tfidf_pack_line[1]

    doc_sentences = sentences
    doc_lines = lines

    return {
        "pages_lines": pages_lines,
        "index": idx,
        "sent_emb": sent_emb,
        "line_emb": line_emb,
        "doc_sentences": doc_sentences,
        "doc_lines": doc_lines,
        "tfidf_pack_sent": tfidf_pack_sent if model is None else None,
        "tfidf_pack_line": tfidf_pack_line if model is None else None,
    }

data = parse_and_index(pdf_bytes)

tabs = st.tabs(["🔎 Ask the Judgment", "📍 Find Page & Line", "📝 Summary"])

# ---------------------------------
# TAB 1 — Ask the Judgment
# ---------------------------------
with tabs[0]:
    st.subheader("Ask if the judgment addresses your point")
    q = st.text_input("Type a sentence/issue to check (e.g., “Whether notice under Section 138 was duly served”).")
    sens = st.slider("Strictness (higher = stricter match)", 0.2, 0.95, 0.55, 0.05)
    k_hits = st.slider("How many supporting snippets to show", 1, 8, 5)

    if q:
        q_vec = embed_query(q, model=model, tfidf_pack=data["tfidf_pack_sent"])
        if model is not None:
            sim = np.dot(q_vec, data["sent_emb"].T).ravel()
        else:
            sim = cosine_similarity(q_vec, data["sent_emb"]).ravel()

        # Decide "is it addressed?"
        best = float(sim.max()) if sim.size else 0.0
        addressed = best >= sens

        st.markdown(f"**Answer:** {'✅ Yes, likely addressed' if addressed else '❌ Not clearly addressed'}  \n"
                    f"(best similarity score: {best:.2f} | threshold: {sens:.2f})")

        # Show top-k snippets with page/line spans
        top_idx = np.argsort(-sim)[:k_hits]
        st.markdown("---")
        st.markdown("**Closest supporting excerpts:**")
        for rank, i in enumerate(top_idx, start=1):
            sent, page, ls, le = data["index"]["sentence_index"][i]
            st.markdown(f"**{rank}. Page {page}, lines {ls}–{le}**  \n> {sent}")

# ---------------------------------
# TAB 2 — Find Page & Line
# ---------------------------------
with tabs[1]:
    st.subheader("Locate exact page and line for a given quote or statement")
    q2 = st.text_input("Paste the specific line/sentence you want to locate exactly.")
    k_lines = st.slider("How many closest lines to show", 1, 10, 5)
    if q2:
        q_vec = embed_query(q2, model=model, tfidf_pack=data["tfidf_pack_line"])
        if model is not None:
            sim = np.dot(q_vec, data["line_emb"].T).ravel()
        else:
            sim = cosine_similarity(q_vec, data["line_emb"]).ravel()
        top_idx = np.argsort(-sim)[:k_lines]

        st.markdown("**Closest line matches:**")
        for rank, i in enumerate(top_idx, start=1):
            page, line_no, text = data["index"]["line_index"][i]
            st.markdown(f"**{rank}. Page {page}, line {line_no}**  \n> {text}")

# ---------------------------------
# TAB 3 — Summary
# ---------------------------------
with tabs[2]:
    st.subheader("Get a concise summary (1–2 paragraphs)")
    max_sents = st.slider("Maximum sentences to include", 3, 12, 6)
    if st.button("Generate Summary"):
        sents = [s for (s, p, ls, le) in data["index"]["sentence_index"]]
        summ = extractive_summary(sents, max_sents=max_sents)
        if not summ:
            st.warning("Could not build a summary from this PDF.")
        else:
            st.write(summ)
